agent/lib_aws_sdk_php.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Functions relating to instrumenting the AWS-SDK-PHP.
+ * https://github.com/aws/aws-sdk-php
+ */
+#include "php_agent.h"
+#include "php_wrapper.h"
+#include "fw_hooks.h"
+#include "fw_support.h"
+#include "util_logging.h"
+#include "php_call.h"
+#include "lib_aws_sdk_php.h"
+
+#define PHP_PACKAGE_NAME "aws/aws-sdk-php"
+#define PHP_AWS_CLASS_PREFIX "Supportability/library/aws/aws-sdk-php/"
+
+/*
+ * In a normal course of events, the following line will always work
+ * zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version")
+ * By the time we have detected the existence of the aws-sdk-php and with
+ * default composer profject settings, it callable even from
+ * nr_aws_sdk_php_enable which will automatically load the class if it isn't
+ * loaded yet and then evaluate the string. However, in the rare case that files
+ * are not loaded via autoloader and/or have non-default composer classload
+ * settings, if the class is not found, PHP 8.2+ will generate a fatal
+ * unrecoverable uncatchable error error whenever it cannot find a class. While
+ * calling this from nr_aws_sdk_php_enable would have been great and would allow
+ * the sdk version value to be set only once, to avoid the very unlikely but not
+ * impossible fatal error, this will be called from the
+ * "Aws\\ClientResolver::_apply_user_agent" wrapper which GUARANTEES that
+ * aws/sdk exists and is already loaded.
+ *
+ *
+ * Additionally given that aws-sdk-php is currently detected from the
+ * AwsClient.php file, this method will always be called when a client is
+ * created. Having separate functionality to extract from Aws/Sdk::__constructor
+ * is both not required and is redundant and causes additional overhead.
+ */
+extern void lib_aws_sdk_php_handle_version() {
+  char* version = NULL;
+  zval retval;
+  int result = FAILURE;
+
+  result = zend_eval_string("Aws\\Sdk::VERSION;", &retval, "Get AWS Version");
+
+  // Add php package to transaction
+  if (SUCCESS == result) {
+    if (nr_php_is_zval_non_empty_string(&retval)) {
+      version = Z_STRVAL(retval);
+    }
+    zval_dtor(&retval);
+  }
+
+  /* Use the Aws\Sdk::VERSION to determine the version */
+  if (NRINI(vulnerability_management_package_detection_enabled)) {
+    /* Add php package to transaction */
+    nr_txn_add_php_package(NRPRG(txn), PHP_PACKAGE_NAME, version);
+  }
+  nr_fw_support_add_package_supportability_metric(NRPRG(txn), PHP_PACKAGE_NAME,
+                                                  version);
+}
+
+extern void lib_aws_sdk_php_add_supportability_metric(const char* metric_name) {
```

Comment:
I am concerned about the performance of this function. It seemingly copies the name twice. Once into the object array during the split, and then once into the buffer. And this is executed on every single call to specific functions.

Reply:
It operates only on calls to the Client's constructor, so theoretically shouldn't happen often, but optimization is always the goal.

options are:

1) We could either accept the `\\` in the name
2) instead of strplit we could iterate through the string and manually remove/replace.

Checkout commit for option 2 here: [7d774ef](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/7d774ef904c4da10b012c89cc46fce6041dc8d21)

Reply:
I like option 2. Maybe we could extract the code into an `nr_strreplace` function?

Reply:
Something else interesting from the commit above: The modification is in-place on the `wraprec->classname` field. If we determine that this doesn't have negative side-effects, this string modification would only need to happen once during wrapper creation. If, however, there are side effects, we probably want to do a single copy each time this is called.

Reply:
Modifying in place wouldn't work as we use that field for other things which was why manipulation was only happening on the copied version in the buffer.

After considering and testing more, I think I overthought this and an solving a problem that doesn't exist.
The metric naming details in agent-specs don't have anything against backslashes, and the backend properly shows it as `Supportability/library/aws/aws-sdk-php/Aws\S3\S3Client`

Leaving it as-is (without replacing the backslash) seems the most efficient.

After experimentation, option one seems better in this case.  It let's us simplify code readability,  reduce string copy/manipulation, and minimize the effect of enabling supportability metrics.

<img width="466" alt="image" src="https://github.com/user-attachments/assets/45c42e71-d3aa-48fc-bcd7-890fd2a9ee6a">

Please see [1148df3](https://github.com/newrelic/newrelic-php-agent/pull/932/commits/1148df3439920382b561b616ba412a7b3a023c08)